File: tests/migrate.test.ts
import { describe, expect, test } from 'bun:test';
import { mkdtempSync, readFileSync, writeFileSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join } from 'node:path';
import { loadLibraryMigrating, migrateLibrary, parseLibraryV1, rekeyState } from '../src/migrate';
import { LibraryError } from '../src/library';

// v1 → v2 migration: fresh UUIDs, a name→id map that also rekeys state, and an
// atomic auto-on-load upgrade that leaves the v1 file intact on failure.

const dir = () => mkdtempSync(join(tmpdir(), 'potato-mig-'));

describe('migrateLibrary', () => {
  test('turns name-keyed v1 commands into a v2 array with fresh ids, preserving order', () => {
    const v1 = {
      version: 1 as const,
      commands: {
        deploy: { command: 'ssh prod', description: 'roll out' },
        list: { command: 'ls' },
      },
    };
    const { library, nameToId } = migrateLibrary(v1);
    expect(library.version).toBe(2);
    expect(library.commands.map((c) => c.name)).toEqual(['deploy', 'list']);
    expect(library.commands[0]!.command).toBe('ssh prod');
    expect(library.commands[0]!.description).toBe('roll out');
    expect(library.commands[1]!.description).toBeUndefined();
    // ids minted and threaded into the map
    expect(nameToId.get('deploy')).toBe(library.commands[0]!.id);
    expect(library.commands[0]!.id).not.toBe(library.commands[1]!.id);
  });

  test('preserves unknown fields at top level and per entry', () => {
    const v1 = { version: 1 as const, color: 'blue', commands: { x: { command: 'ls', note: 'keep' } } };
    const { library } = migrateLibrary(v1);
    expect((library as { color?: string }).color).toBe('blue');
    expect((library.commands[0] as { note?: string }).note).toBe('keep');
  });
});

describe('rekeyState', () => {
  test('moves entries from name keys to id keys and drops orphans', () => {
    const map = new Map([['deploy', 'id-1']]);
    const state = {
      deploy: { lastUsedAt: '2026-07-01T00:00:00.000Z' },
      gone: { lastUsedAt: '2026-01-01T00:00:00.000Z' },
    };
    expect(rekeyState(state, map)).toEqual({ 'id-1': { lastUsedAt: '2026-07-01T00:00:00.000Z' } });
  });
});

describe('parseLibraryV1', () => {
  test('fails loud on a malformed v1 file, naming the source', () => {
    expect(() => parseLibraryV1('{ nope', 'commands.json')).toThrow(LibraryError);
    expect(() => parseLibraryV1('{"version":1,"commands":{"x":{}}}', 'commands.json')).toThrow(/commands\.json/);
  });
});

describe('loadLibraryMigrating', () => {
  test('a missing library is an empty v2 library, not migrated', () => {
    const home = dir();
    const result = loadLibraryMigrating(join(home, 'commands.json'), join(home, 'state.json'));
    expect(result).toEqual({ library: { version: 2, commands: [] }, state: {}, migrated: false });
  });

  test('a v2 library loads unchanged, not migrated', () => {
    const home = dir();
    const cmds = join(home, 'commands.json');
    writeFileSync(cmds, JSON.stringify({ version: 2, commands: [{ id: 'a', name: 'x', command: 'ls' }] }));
    const result = loadLibraryMigrating(cmds, join(home, 'state.json'));
    expect(result.migrated).toBe(false);
    expect(result.library.commands[0]!.name).toBe('x');
  });

  test('a v1 library migrates, writes v2 back, and rekeys state', () => {
    const home = dir();
    const cmds = join(home, 'commands.json');
    const st = join(home, 'state.json');
    writeFileSync(cmds, JSON.stringify({ version: 1, commands: { deploy: { command: 'ssh prod' } } }));
    writeFileSync(st, JSON.stringify({ deploy: { lastUsedAt: '2026-07-01T00:00:00.000Z' } }));

    const result = loadLibraryMigrating(cmds, st);
    expect(result.migrated).toBe(true);
    const id = result.library.commands[0]!.id;
    // file written back as v2
    expect(JSON.parse(readFileSync(cmds, 'utf8')).version).toBe(2);
    // state rekeyed to the new id, on disk and in the result
    expect(result.state[id]!.lastUsedAt).toBe('2026-07-01T00:00:00.000Z');
    expect(JSON.parse(readFileSync(st, 'utf8'))[id].lastUsedAt).toBe('2026-07-01T00:00:00.000Z');
    // a second load sees v2 and does not migrate again (idempotent)
    expect(loadLibraryMigrating(cmds, st).migrated).toBe(false);
  });

  test('a malformed v1 library fails loud, leaving the file untouched', () => {
    const home = dir();
    const cmds = join(home, 'commands.json');
    const original = JSON.stringify({ version: 1, commands: { x: { command: '' } } });
    writeFileSync(cmds, original);
    expect(() => loadLibraryMigrating(cmds, join(home, 'state.json'))).toThrow(LibraryError);
    expect(readFileSync(cmds, 'utf8')).toBe(original);
  });
});

File: internal/migrate/migrate_test.go
package migrate

import (
	"os"
	"path/filepath"
	"strings"
	"testing"

	"github.com/luojiahai/potato/internal/library"
	"github.com/luojiahai/potato/internal/state"
)

// v1 → v2 migration: fresh UUIDs, a name→id map that also rekeys state, and an
// atomic auto-on-load upgrade that leaves the v1 file intact on failure.

func parseV1(t *testing.T, text string) LibraryV1 {
	t.Helper()
	v1, err := ParseV1(text, "commands.json")
	if err != nil {
		t.Fatalf("ParseV1: %v", err)
	}
	return v1
}

func TestMigratePreservesOrderAndMintsIDs(t *testing.T) {
	v1 := parseV1(t, `{"version":1,"commands":{"deploy":{"command":"ssh prod","description":"roll out"},"list":{"command":"ls"}}}`)
	lib, nameToID := Migrate(v1)

	if lib.Version != 2 {
		t.Errorf("version = %d", lib.Version)
	}
	if len(lib.Commands) != 2 || lib.Commands[0].Name != "deploy" || lib.Commands[1].Name != "list" {
		t.Fatalf("order not preserved: %+v", lib.Commands)
	}
	if lib.Commands[0].Command != "ssh prod" {
		t.Errorf("command = %q", lib.Commands[0].Command)
	}
	if lib.Commands[0].Description == nil || *lib.Commands[0].Description != "roll out" {
		t.Errorf("description = %v", lib.Commands[0].Description)
	}
	if lib.Commands[1].Description != nil {
		t.Errorf("description should be absent, got %v", *lib.Commands[1].Description)
	}
	if nameToID["deploy"] != lib.Commands[0].ID {
		t.Error("name→id map does not match the minted id")
	}
	if lib.Commands[0].ID == lib.Commands[1].ID {
		t.Error("ids are not unique")
	}
}

func TestMigratePreservesUnknownFields(t *testing.T) {
	v1 := parseV1(t, `{"version":1,"color":"blue","commands":{"x":{"command":"ls","note":"keep"}}}`)
	lib, _ := Migrate(v1)
	if string(lib.Extra["color"]) != `"blue"` {
		t.Errorf("top-level unknown field = %s", lib.Extra["color"])
	}
	if string(lib.Commands[0].Extra["note"]) != `"keep"` {
		t.Errorf("entry unknown field = %s", lib.Commands[0].Extra["note"])
	}
}

// v1 tolerated arbitrary unknown fields; reserved-looking ones must not
// override the minted id, the map-keyed name, or the command.
func TestMigrateStrayReservedFieldsCannotClobberIdentity(t *testing.T) {
	v1 := parseV1(t, `{"version":1,"commands":{"a":{"command":"echo a","id":"evil","name":"spoof"},"b":{"command":"echo b","id":"evil"}}}`)
	lib, nameToID := Migrate(v1)
	a, b := lib.Commands[0], lib.Commands[1]

	if a.ID == "evil" || a.ID == b.ID {
		t.Errorf("minted ids did not win: %q %q", a.ID, b.ID)
	}
	if nameToID["a"] != a.ID {
		t.Error("name→id map does not match")
	}
	if a.Name != "a" || a.Command != "echo a" {
		t.Errorf("name/command clobbered: %q %q", a.Name, a.Command)
	}
	if _, err := library.Parse(library.Serialize(lib), "migrated"); err != nil {
		t.Errorf("migrated library does not round-trip: %v", err)
	}
}

func TestRekeyStateDropsOrphans(t *testing.T) {
	s := state.State{"deploy": {LastUsedAt: "2026-01-01T00:00:00Z"}, "gone": {LastUsedAt: "x"}}
	next := RekeyState(s, map[string]string{"deploy": "id-1"})
	if next["id-1"].LastUsedAt != "2026-01-01T00:00:00Z" {
		t.Errorf("state not rekeyed: %v", next)
	}
	if len(next) != 1 {
		t.Errorf("orphan kept: %v", next)
	}
}

func TestLoadMigratesV1OnDisk(t *testing.T) {
	dir := t.TempDir()
	commands := filepath.Join(dir, "commands.json")
	statePath := filepath.Join(dir, "state.json")
	if err := os.WriteFile(commands, []byte(`{"version":1,"commands":{"deploy":{"command":"ssh prod"}}}`), 0o644); err != nil {
		t.Fatal(err)
	}
	if err := os.WriteFile(statePath, []byte(`{"deploy":{"lastUsedAt":"2026-07-01T00:00:00Z"}}`), 0o644); err != nil {
		t.Fatal(err)
	}

	result, err := Load(commands, statePath)
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	if !result.Migrated {
		t.Error("Migrated flag not set")
	}
	id := result.Library.Commands[0].ID
	if result.State[id].LastUsedAt != "2026-07-01T00:00:00Z" {
		t.Errorf("state was not rekeyed onto the new id: %v", result.State)
	}

	// the file on disk is now v2 and parses strictly
	raw, err := os.ReadFile(commands)
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(string(raw), `"version": 2`) {
		t.Errorf("file was not upgraded:\n%s", raw)
	}
	if _, err := library.Parse(string(raw), commands); err != nil {
		t.Errorf("upgraded file does not parse: %v", err)
	}
}

func TestLoadLeavesV2Alone(t *testing.T) {
	dir := t.TempDir()
	commands := filepath.Join(dir, "commands.json")
	if err := os.WriteFile(commands, []byte(`{"version":2,"commands":[{"id":"a","name":"x","command":"ls"}]}`), 0o644); err != nil {
		t.Fatal(err)
	}
	result, err := Load(commands, filepath.Join(dir, "state.json"))
	if err != nil {
		t.Fatalf("Load: %v", err)
	}
	if result.Migrated {
		t.Error("a v2 library should not report a migration")
	}
	if result.Library.Commands[0].ID != "a" {
		t.Error("v2 ids must be left alone")
	}
}

func TestLoadFailsLoudOnAFutureVersion(t *testing.T) {
	dir := t.TempDir()
	commands := filepath.Join(dir, "commands.json")
	if err := os.WriteFile(commands, []byte(`{"version":3,"commands":[]}`), 0o644); err != nil {
		t.Fatal(err)
	}
	if _, err := Load(commands, filepath.Join(dir, "state.json")); err == nil {
		t.Error("expected a failure on an unknown version")
	}
}
